File: backend/src/routers/images.py
"""
图片路由
(CRUD /images/*)
"""

import os

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.database import get_db
from src.models.user import User
from src.routers.users import get_current_user
from src.schemas.ai_task import AITaskSubmitResponse
from src.schemas.image import (
    AIEditRequest,
    EditImageRequest,
    ImageListResponse,
    ImageResponse,
    ImageUploadResponse,
    UpdateImageNameRequest,
)
from src.services import image_service, task_service

router = APIRouter(prefix="/api/images", tags = ["图片"])

class URLUploadRequest(BaseModel):
    url: str
    custom_name: str | None = None

@router.post("/upload", response_model=ImageUploadResponse)
async def upload(
    file: UploadFile = File(...),
    custom_name: str = Form(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """上传文件（可选自定义名称）"""
    image = await image_service.upload_file(db, file, current_user, custom_name)
    return _build_upload_response(image)

@router.post("/upload-url", response_model=ImageUploadResponse)
async def upload_url(
    req: URLUploadRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """上传URL图片"""
    image = await image_service.upload_from_url(db, req.url, current_user, req.custom_name)
    return _build_upload_response(image)

@router.get("", response_model=ImageListResponse)
async def list_images(
    skip: int = Query(0,ge = 0),    # 跳过前 N 条记录，ge=0 表示最小值为 0
    limit: int = Query(20, ge = 1, le = 100),    # 最多返回 N 条记录，最小 1，最大 100
    search: str | None = Query(None, description="搜索关键词（匹配图片名称）"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """获取图片列表（支持搜索 + Redis 缓存）"""
    return await image_service.get_user_images(db, current_user, skip, limit, search)

@router.get("/{image_id}", response_model=ImageResponse)
def get_image(
    image_id : int,
    db : Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """获取图片详情"""
    return image_service.get_image_detail(db, image_id, current_user)

@router.get("/{image_id}/download")
def download_original(
    image_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """下载原始格式图片"""
    image = image_service.get_image_detail(db, image_id, current_user)
    if not os.path.exists(image.file_path):
        raise HTTPException(status_code=404, detail="原始图片不存在")
    return FileResponse(
        image.file_path,
        filename=image.original_name,
        media_type="application/octet-stream",
    )

@router.delete("/{image_id}")
async def delete_image(
    image_id : int,
    db : Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """删除图片"""
    image_service.delete_image(db, image_id, current_user)
    return {"message": "图片删除成功"}


@router.patch("/{image_id}/name", response_model=ImageResponse)
async def rename_image(
    image_id: int,
    req: UpdateImageNameRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """修改图片名称"""
    return image_service.update_image_name(db, image_id, req.custom_name, current_user)


@router.post("/{image_id}/edit", response_model=ImageUploadResponse)
async def edit_image(
    image_id: int,
    req: EditImageRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """编辑图片（裁剪/旋转/翻转），支持覆盖或另存"""
    image = await image_service.edit_image(
        db, image_id, req.operations, req.save_mode, req.custom_name, current_user
    )
    return _build_upload_response(image)


@router.post("/{image_id}/replace", response_model=ImageUploadResponse)
async def replace_image(
    image_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """用新图片文件覆盖原图（AI 抠图/区域编辑结果）"""
    image = await image_service.replace_image(db, image_id, file, current_user)
    return _build_upload_response(image)

@router.post("/{image_id}/ai-edit", response_model=AITaskSubmitResponse)
async def ai_edit_image(
    image_id: int,
    req: AIEditRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """AI 区域编辑（涂鸦 + 指令）：提交即返回 task_id，前端轮询 /api/tasks/{task_id} 取状态与结果"""
    task = task_service.create_ai_edit_task(
        db, image_id, req.prompt, req.image_base64, current_user, req.color_name
    )
    task_service.schedule_ai_edit_task(task.id)
    return AITaskSubmitResponse(task_id=task.id, status=task.status)

def _build_upload_response(image) -> dict:
    """构建上传响应(含URL)"""
    return {
        "id": image.id,
        "original_name": image.original_name,
        "custom_name": image.custom_name,
        "display_name": image.display_name,
        "file_size": image.file_size,
        "width": image.width,
        "height": image.height,
        "image_url": image.image_url,
        "thumbnail_url": image.thumbnail_url,
    }

File: frontend/src/api/images.ts

import api from "./client"
import type { AITaskSubmitResponse } from "./tasks";

// interface 定义了每个接口的请求/响应结构，
// 和 pydantic.BaseModel 是同样的概念——前端也要"校验"数据结构。
export interface ImageItem {
    id: number
    filename: string
    original_name: string
    custom_name: string | null
    display_name: string 
    file_size: number
    mime_type: string
    width: number | null
    height: number | null
    created_at: string
    thumbnail_url: string | null
    image_url: string | null
    // 阶段 21：个人标签随列表一起返回（卡片展示 + 点击筛选）
    tags: string[]
}

export interface ImageListResponse {
    total: number
    items: ImageItem[]
}

// 阶段 21：当前用户的标签统计（图库筛选条数据源，按使用次数降序）
export interface ImageTagStat {
    name: string
    count: number
}

export interface ImageTagListResponse {
    total: number
    items: ImageTagStat[]
}

export interface ImageUploadRequest {
    id: number
    original_name: string
    custom_name: string | null
    display_name: string
    file_size: number
    width: number
    height: number
    image_url: string
    thumbnail_url: string
}

// 获取图片列表（阶段 21：新增 tag 精确筛选，与 search 同时给出取 AND）
export function getImages(skip = 0, limit = 20, search?: string, tag?: string): Promise<ImageListResponse> {
    const params: Record<string, any> = { skip, limit }
    if(search) params.search = search
    if(tag) params.tag = tag
    return api.get("/images", { params: params })
}

// 获取当前用户的标签统计（阶段 21：图库筛选条）
export function getImageTags(): Promise<ImageTagListResponse> {
    return api.get("/images/tags")
}

// 获取图片详情（含个人图库标签）
export function getImageDetail(imageId: number): Promise<ImageItem> {
    return api.get(`/images/${imageId}`)
}

// 上传图片文件
export function uploadImage(file: File, customName?: string): Promise<ImageUploadRequest> {
    const formData = new FormData()
    formData.append("file", file)
    if(customName) formData.append("custom_name", customName)
    return api.post("/images/upload", formData, {
        headers: {
            "Content-Type": "multipart/form-data"
        }
    })
}

// URL上传
export function uploadImageByUrl(url: string, customName?: string): Promise<ImageUploadRequest> {
    return api.post("/images/upload-url", { url, custom_name: customName || null })
}

// 删除图片
export function deleteImage(imageId: number): Promise<{ message: string }> {
    return api.delete(`/images/${imageId}`)
}

// 修改图片名称
export function updateImageName(imageId: number, customName: string): Promise<ImageItem> {
    return api.patch(`/images/${imageId}/name`, { custom_name: customName })
}

// 添加个人图库标签（阶段 19：只影响个人标签，已提交的公开标签不受影响）
export function addImageTags(imageId: number, tags: string[]): Promise<{ message: string; tags: string[] }> {
    return api.post(`/images/${imageId}/tags`, { tags })
}

// 移除个人图库标签
export function removeImageTag(imageId: number, tagName: string): Promise<{ message: string }> {
    return api.delete(`/images/${imageId}/tags/${encodeURIComponent(tagName)}`)
}

// 下载原图URL（直接打开即可触发浏览器下载）
export function getDownloadUrl(imageId: number): string {
    return `/api/images/${imageId}/download`
}

// 下载原图（Axios携带Token，以blob方式接收）
export async function downloadOriginalImage(imageId: number, filename: string): Promise<void> {
    const response = await api.get(`/images/${imageId}/download`, { responseType: "blob", } as any)
    //创建临时URL触发浏览器下载
    const url = window.URL.createObjectURL(new Blob ([response as any]))
    const link = document.createElement("a")
    link.href = url
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)
}

// 编辑操作类型
export interface EditOperation {
    type: "rotate" | "flip" | "crop"
    angle?: number          // rotate: 90/180/270
    direction?: string      // flip: "horizontal"/"vertical"
    left?: number           // crop
    top?: number
    right?: number
    bottom?: number
}

// 编辑图片（裁剪/旋转/翻转）
export function editImage(
    imageId: number,
    operations: EditOperation[],
    saveMode: "overwrite" | "new",
    customName?: string
): Promise<ImageUploadRequest> {
    return api.post(`/images/${imageId}/edit`, {
        operations,
        save_mode: saveMode,
        custom_name: customName || null
    })
}

// 替换图片（覆盖原图，用于 AI 抠图结果）
export function replaceImage(imageId: number, file: Blob): Promise<ImageUploadRequest> {
    const formData = new FormData()
    // Blob 直接 append 会丢失文件名（后端拿到 "blob"），无法识别扩展名
    const namedFile = new File([file], "result.png", { type: file.type || "image/png" })
    formData.append("file", namedFile)
    return api.post(`/images/${imageId}/replace`, formData, {
        headers: { "Content-Type": "multipart/form-data" }
    })
}

// AI 区域编辑（涂鸦 + 指令）：提交后立即返回 task_id，结果由 /api/tasks 轮询获取
// colorName 用于后端拼接 prompt
export function aiEditImage(
    imageId: number,
    prompt: string,
    imageBase64: string,
    colorName = "红色"
): Promise<AITaskSubmitResponse> {
    return api.post(`/images/${imageId}/ai-edit`, { prompt, image_base64: imageBase64, color_name: colorName })
}
